hypershift-operator/controllers/nodepool/kubevirt/kubevirt.go

```diff
@@ -1,47 +1,68 @@
 	hyperv1 "github.com/openshift/hypershift/api/v1beta1"
 	"github.com/openshift/hypershift/support/releaseinfo"
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
 	apiresource "k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
 	kubevirtv1 "kubevirt.io/api/core/v1"
 	"kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
 	capikubevirt "sigs.k8s.io/cluster-api-provider-kubevirt/api/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func defaultImage(releaseImage *releaseinfo.ReleaseImage) (string, error) {
+const (
+	containerImagePrefix = "docker://"
+
+	// a CDI annotation for DataVolume, to not wait to first customer, but start importing immediately.
+	cdiImmediateBindingAnnotation = "cdi.kubevirt.io/storage.bind.immediate.requested"
+
+	bootImagePVCAnnotationImageHash = "hypershift.openshift.io/kubevirt-boot-image-hash" // temp value
+	bootImagePVCPrefix              = "kubevirt-boot-image-"
+)
+
+func defaultImage(releaseImage *releaseinfo.ReleaseImage) (string, string, error) {
 	arch, foundArch := releaseImage.StreamMetadata.Architectures["x86_64"]
 	if !foundArch {
-		return "", fmt.Errorf("couldn't find OS metadata for architecture %q", "x64_64")
+		return "", "", fmt.Errorf("couldn't find OS metadata for architecture %q", "x64_64")
 	}
 	openStack, exists := arch.Artifacts["openstack"]
 	if !exists {
-		return "", fmt.Errorf("couldn't find OS metadata for openstack")
+		return "", "", fmt.Errorf("couldn't find OS metadata for openstack")
 	}
 	artifact, exists := openStack.Formats["qcow2.gz"]
 	if !exists {
-		return "", fmt.Errorf("couldn't find OS metadata for openstack qcow2.gz")
+		return "", "", fmt.Errorf("couldn't find OS metadata for openstack qcow2.gz")
 	}
 	disk, exists := artifact["disk"]
 	if !exists {
-		return "", fmt.Errorf("couldn't find OS metadata for the openstack qcow2.gz disk")
+		return "", "", fmt.Errorf("couldn't find OS metadata for the openstack qcow2.gz disk")
 	}
 
-	return disk.Location, nil
+	return disk.Location, disk.SHA256, nil
 }
 
-func GetImage(nodePool *hyperv1.NodePool, releaseImage *releaseinfo.ReleaseImage) (string, error) {
+func GetImage(ctx context.Context, nodePool *hyperv1.NodePool, releaseImage *releaseinfo.ReleaseImage) (string, string, error) {
 	if nodePool.Spec.Platform.Kubevirt != nil &&
 		nodePool.Spec.Platform.Kubevirt.RootVolume != nil &&
 		nodePool.Spec.Platform.Kubevirt.RootVolume.Image != nil &&
 		nodePool.Spec.Platform.Kubevirt.RootVolume.Image.ContainerDiskImage != nil {
 
-		return fmt.Sprintf("docker://%s", *nodePool.Spec.Platform.Kubevirt.RootVolume.Image.ContainerDiskImage), nil
+		imageName := *nodePool.Spec.Platform.Kubevirt.RootVolume.Image.ContainerDiskImage
+		imageHash, err := getImageDigest(ctx, imageName)
+		if err != nil {
+			return "", "", fmt.Errorf("failed to get the image hash; %w", err)
+		}
+
+		return containerImagePrefix + imageName, imageHash, nil
```

Comment:
for now, i'd suggest only caching images provided by the release payload, where the sha is known beforehand.